File: jva/src/clientAis/networking/CommunicationHandler.java
package clientAis.networking;

import java.io.*;
import java.net.Socket;
import java.util.Arrays;
import java.util.stream.IntStream;

public class CommunicationHandler implements Closeable {

    public boolean isClosed;

    private static final int BUFFER_SIZE = 65535;

    private final Socket socket;
    private final ConnectionCredentials serverCredentials;
    private final char[] bufferData;
    private BufferedReader reader;
    private final PrintWriter writer;

    public CommunicationHandler(Socket socket, ConnectionCredentials serverCredentials) {
        this.socket = socket;
        this.serverCredentials = serverCredentials;
        this.isClosed = false;
        this.bufferData = new char[BUFFER_SIZE];
        try {
            this.reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            this.writer = new PrintWriter(socket.getOutputStream(), true);
        }
        catch (IOException ioException) {
            throw new RuntimeException(ioException);
        }
    }

    @Override
    public void close() {
        isClosed = true;
    }

    public void send(String communicationData) {
        raiseExceptionIfDisconnectedFromServer("send");
        writer.write(communicationData);
        writer.flush();
    }

    public String receive() {
        raiseExceptionIfDisconnectedFromServer("receive");
        try {
            final int lengthRead = reader.read(bufferData);
            if(lengthRead == -1) {
                close();
                throw new RuntimeException("Remote socket has been shutdown while receiving from it");
            }
            final char[] copiedBufferData = Arrays.copyOfRange(bufferData, 0, lengthRead);
            return new String(copiedBufferData);
        }
        catch (IOException ioException) {
            ioException.printStackTrace();
            close();
            if(socket.isClosed()) {
                throw new RuntimeException("Remote socket has been shutdown while receiving from it");
            }
            try {
                reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            }
            catch (IOException ioException2) {
                ioException2.printStackTrace();
                throw new RuntimeException("Remote socket has been shutdown while receiving from it");
            }
            return receive();
        }
    }

    private void raiseExceptionIfDisconnectedFromServer(String nameOfAction) {
        if(isClosed) {
            throw new RuntimeException("Cannot " + nameOfAction + " data through a disconnected socket.");
        }
    }
}

File: py/clientAis/networking/CommunicationHandler.py
import socket
from typing import Tuple

class CommunicationHandler:
    def __init__(self, sock: socket.socket, endpoint: Tuple[str, int]):
        self.sock = sock
        self.endpoint = endpoint
        self.closed = False
        self.BUFFER_SIZE = 65535

    def send(self, communication: str):
        if self.closed:
            raise Exception("Cannot send data through a disconnected socket.")

        try:
            return self.sock.send(communication.encode("ascii"))
        except:
            self.closed = True

    def receive(self) -> str:
        if self.closed:
            raise Exception("Cannot receive data through a disconnected socket.")

        try:
            communication = self.sock.recv(self.BUFFER_SIZE).decode("ascii")

            if len(communication) == 0:
                self.closed = True
                raise Exception("Remote socket has been shutdown while receiving from it")

            return communication

        except:
            self.closed = True
            raise

    def close(self):
        if self.closed:
            return

        self.closed = True
        return self.sock.close()
